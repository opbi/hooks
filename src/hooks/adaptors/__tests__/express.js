import adaptorExpress from '../express';

import eventLogger from '../../event-logger';

const resMock = {
  send: jest.fn(),
};
const nextMock = jest.fn();
const loggerMock = {
  info: jest.fn(),
  error: jest.fn(),
};

describe('adaptorExpress', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('has default pattern to adapt action to express function', async () => {
    const handlerAction = ({ count }, meta, { res }) => {
      res.send(count);
    };

    const handler = adaptorExpress()(handlerAction);

    const req = {
      contentType: 'json',
      meta: {
        requestId: '57cec994-ea1e-4555-81d6-45de628c3af9',
      },
      body: {
        count: 1,
      },
      logger: loggerMock,
    };

    await handler(req, resMock, nextMock);

    expect(resMock.send.mock.calls).toMatchSnapshot();
  });

  it('can add additional adaption via config', async () => {
    const validateAction = ({ count }, meta, { logger, next }) => {
      if (count) {
        logger.info('validation passed', meta);
        return next();
      }
      return next(Error('invalid input - count not found'));
    };

    const validateMiddleware = adaptorExpress({
      context: (req) => ({ logger: req.logger }),
    })(validateAction);

    const req = {
      contentType: 'json',
      meta: {
        requestId: '57cec994-ea1e-4555-81d6-45de628c3af9',
      },
      body: {
        count: 1,
      },
      logger: loggerMock,
    };

    await validateMiddleware(req, resMock, nextMock);

    expect(nextMock.mock.calls).toMatchSnapshot();
    expect(loggerMock.info.mock.calls).toMatchSnapshot();
  });

  it('chainable with hooks', async () => {
    const handlerAction = ({ count }, meta, { res }) => {
      res.send(count);
    };

    const adaptorExpressWithLogger = adaptorExpress({
      context: (req) => ({ logger: req.logger }),
    });
    const handler = adaptorExpressWithLogger(eventLogger()(handlerAction));

    const req = {
      contentType: 'json',
      meta: {
        requestId: '57cec994-ea1e-4555-81d6-45de628c3af9',
      },
      body: {
        count: 1,
      },
      logger: loggerMock,
    };

    await handler(req, resMock, nextMock);

    expect(resMock.send.mock.calls).toMatchSnapshot();
    expect(loggerMock.info.mock.calls).toMatchSnapshot();
  });
});
