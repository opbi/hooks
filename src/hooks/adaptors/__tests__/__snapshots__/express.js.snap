// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adaptorExpress can add additional adaption via config 1`] = `
Array [
  Array [],
]
`;

exports[`adaptorExpress can add additional adaption via config 2`] = `
Array [
  Array [
    "validation passed",
    Object {
      "requestId": "57cec994-ea1e-4555-81d6-45de628c3af9",
    },
  ],
]
`;

exports[`adaptorExpress chainable with hooks 1`] = `
Array [
  Array [
    1,
  ],
]
`;

exports[`adaptorExpress chainable with hooks 2`] = `
Array [
  Array [
    Object {
      "event": "handlerAction",
      "requestId": "57cec994-ea1e-4555-81d6-45de628c3af9",
    },
    "handlerAction",
  ],
]
`;

exports[`adaptorExpress has default pattern to adapt action to express function 1`] = `
Array [
  Array [
    1,
  ],
]
`;
